### PR TITLE
[zero] hybrid adam and place part of OS on cuda

### PR DESCRIPTION
**Pull request by feifeibear**

### CODE REVIEW

File: colossalai/zero/sharded_optim/sharded_optim_v2.py [hunk 1]
```diff
@@ -111,23 +111,46 @@ def __init__(self,
...
+        self._first = True
 
     def step(self, *args, **kwargs):
+
+        # When the first time enter the step(), we calculate the margin space for OS
+        # If margin space is enough, part of param fp32 are initialized on GPU.
+        if self._first:
+            self._first = False
+            cuda_margin_space = self.model.cuda_margin_space
+            # div 3 because the space is shared with param fp32, M and V
```
Comment (ver217) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:123]: What about grad fp32?

Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:123]: OK, as we discussed. If we alway keep a grad fp32, it should be div 4 here.

File: colossalai/nn/optimizer/cpu_adam.py [hunk 2]
```diff
@@ -51,47 +71,64 @@ def step(self, closure=None):
...
+                        state['exp_avg'] = torch.zeros_like(p.data, dtype=torch.float, device=cuda_device)
+                        # gradient variances
+                        state['exp_avg_sq'] = torch.zeros_like(p.data, dtype=torch.float, device=cuda_device)
+                        # memory_format=torch.preserve_format)
+
+                    state['step'] += 1
+                    beta1, beta2 = group['betas']
+
+                    # FIXME() prepare grad on cuda
+                    if p.grad.device.type == 'cpu':
```
Comment (Gy-Lu) [inline colossalai/nn/optimizer/cpu_adam.py:119]: Would `p` and `p.grad` have different devices?

File: colossalai/nn/optimizer/cpu_adam.py [hunk 3]
```diff
@@ -51,47 +71,64 @@ def step(self, closure=None):
...
+
+                if p.device.type == 'cpu':
+                    # State initialization
+                    # intended device for step
+                    cpu_device = torch.device('cpu')
+                    if len(state) == 0:
+                        state['step'] = 0
+
+                        # gradient momentums
+                        state['exp_avg'] = torch.zeros_like(p.data, dtype=torch.float, device=cpu_device)
```
[the remark this answers is not in the document]
Comment (feifeibear) [reply colossalai/nn/optimizer/cpu_adam.py:92]: https://pytorch.org/docs/stable/generated/torch.zeros_like.html
you can change the device.

File: colossalai/nn/optimizer/cpu_adam.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/nn/optimizer/cpu_adam.py:119]: Yes, now p.grad is always on cpu.

File: colossalai/zero/sharded_optim/sharded_optim_v2.py (hunk 1 above)
Comment (Gy-Lu) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:123]: ok, I get it.

File: colossalai/nn/optimizer/cpu_adam.py (hunk 2 above)
Comment (Gy-Lu) [reply colossalai/nn/optimizer/cpu_adam.py:119]: ok, I get it.

File: colossalai/nn/optimizer/cpu_adam.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gy-Lu) [reply colossalai/nn/optimizer/cpu_adam.py:92]: I mean, it is unnecessary to specify the device, and you can reduce the redundant code in `if...else` by using the default value.

**Pull request closed without merging**